File: client/dotnet-client/NeoApi/Api/PortfolioApi.cs
using System.Text.Json;
using Kotak.Neo.Rest;
using Kotak.Neo.Validation;

namespace Kotak.Neo.Api;

public class PortfolioApi
{
    private readonly RestClient _rest;
    private readonly Configuration _config;

    public PortfolioApi(RestClient rest, Configuration config)
    {
        _rest = rest;
        _config = config;
    }

    private Dictionary<string, string> Auth(string contentType) => new()
    {
        ["Sid"] = _config.EditSid ?? "",
        ["Auth"] = _config.EditToken ?? "",
        ["Content-Type"] = contentType,
    };

    private Dictionary<string, string> ServerId() =>
        new() { ["sId"] = _config.ServerId ?? "" };

    public async Task<JsonElement> PositionsAsync(CancellationToken ct = default)
    {
        var headers = Auth("application/x-www-form-urlencoded");
        headers["accept"] = "application/json";
        var (_, data, _) = await _rest.RequestAsync("GET",
            _config.GetUrl("positions"),
            queryParams: ServerId(), headers: headers, ct: ct);
        return data;
    }

    public async Task<JsonElement> HoldingsAsync(CancellationToken ct = default)
    {
        var headers = Auth("application/x-www-form-urlencoded");
        headers["accept"] = "*/*";
        var (_, data, _) = await _rest.RequestAsync("GET",
            _config.GetUrl("holdings"),
            queryParams: ServerId(), headers: headers, ct: ct);
        return data;
    }

    public async Task<JsonElement> LimitsAsync(string segment = "ALL", string exchange = "ALL", string product = "ALL", CancellationToken ct = default)
    {
        Validators.ValidateLimits(segment, exchange, product);
        var (_, data, _) = await _rest.RequestAsync("POST",
            _config.GetUrl("limits"),
            queryParams: ServerId(),
            headers: Auth("application/x-www-form-urlencoded"),
            body: new Dictionary<string, object?>
            {
                ["seg"] = segment, ["exch"] = exchange, ["prod"] = product,
            },
            ct: ct);
        return data;
    }
}

File: client/java-client/src/main/java/com/kotak/neo/client/api/PortfolioApi.java
package com.kotak.neo.client.api;

import com.google.gson.JsonElement;
import com.kotak.neo.client.Configuration;
import com.kotak.neo.client.rest.RestClient;
import com.kotak.neo.client.validation.Validators;

import java.util.HashMap;
import java.util.Map;

public class PortfolioApi {
    private final RestClient rest;
    private final Configuration config;

    public PortfolioApi(RestClient rest, Configuration config) {
        this.rest = rest;
        this.config = config;
    }

    private Map<String, String> auth(String ct) {
        Map<String, String> h = new HashMap<>();
        h.put("Sid", config.editSid == null ? "" : config.editSid);
        h.put("Auth", config.editToken == null ? "" : config.editToken);
        h.put("Content-Type", ct);
        return h;
    }

    private Map<String, String> serverId() {
        Map<String, String> q = new HashMap<>();
        q.put("sId", config.serverId == null ? "" : config.serverId);
        return q;
    }

    public JsonElement positions() {
        Map<String, String> h = auth("application/x-www-form-urlencoded");
        h.put("accept", "application/json");
        return rest.request("GET", config.getUrl("positions"), serverId(), h, null).data;
    }

    public JsonElement holdings() {
        Map<String, String> h = auth("application/x-www-form-urlencoded");
        h.put("accept", "*/*");
        return rest.request("GET", config.getUrl("holdings"), serverId(), h, null).data;
    }

    public JsonElement limits(String segment, String exchange, String product) {
        Validators.validateLimits(segment, exchange, product);
        Map<String, Object> body = new HashMap<>();
        body.put("seg", segment);
        body.put("exch", exchange);
        body.put("prod", product);
        return rest.request("POST", config.getUrl("limits"), serverId(),
                auth("application/x-www-form-urlencoded"), body).data;
    }
}
